ADBench/plot_graphs.py

```diff
@@ -188,25 +219,78 @@ def sorting_key_fun(v):
     lines = zip(all_styles, sorted_vals_by_tool)
 
     handles, labels = [], []
-    for ((color, marker), (tool, n_vals, t_vals)) in lines:
-        # Plot results
-        handles += pyplot.plot(n_vals, t_vals, 
-                               marker=marker, color=color, label=utils.format_tool(tool))
-        labels.append(utils.format_tool(tool))
-
-    # Draw black dots
-    max_len = max(map(lambda h: len(utils.get_non_infinite_y(h)), handles))
-    failed = filter(lambda h: len(utils.get_non_infinite_y(h)) < max_len, handles)
-    failed_x, failed_y = [], []
-    for handle in failed:
-        inf_inds = [i for i, y in enumerate(handle.get_ydata()) if y == float("inf")]
-        last_ind = inf_inds[0] - 1 if len(inf_inds) > 0 else -1
-        failed_x.append(handle.get_xdata()[last_ind])
-        failed_y.append(handle.get_ydata()[last_ind])
-        #handles += pyplot.plot(handle.get_xdata()[last_ind], handle.get_ydata()[last_ind], marker="o", color=(0, 0, 0), linestyle="None", label="Crashed/Terminated")
-    if len(failed_x) > 0:
-        handles += tuple(pyplot.plot(failed_x, failed_y, marker="o", color="k", linestyle="None", label=TERMINATED_LABEL))
-        labels += (TERMINATED_LABEL,)
+    violation_x, violation_y = [], []
+    was_violation = False
+    additional = []
+    violation_handle = None
+
+    # Plot results
+    for ((color, marker), (tool, n_vals, t_vals, violations)) in lines:
+        def vals_with_neighbours_and_violation():
+            # Checking neighbours by shifting t_vals requires that
+            # it is in the order of monotonic n_vals
+            assert n_vals == sorted(n_vals)
+            return zip(
```

Comment:
Looks very nice! I really like this code (much more than my version of it)

Reply:
Thanks!  Hopefully you can see how it is possible to use Python iterators to avoid explicit indexing.  Explicit indexing can get very messy and can be hard to understand.